File: python3/utils/binary_tree.py
from __future__ import annotations
from typing import Any, Optional
from collections import deque

null: None = None


class TreeNode:
    def __init__(self, val: Any = 0, left: TreeNode | None = None, right: TreeNode | None = None) -> None:
        self.val = val
        self.left = left
        self.right = right

    def insert(self, val: Any) -> None:
        # already exist => do nothing
        if self.val == val:
            return
        # insert to the left
        elif val < self.val:
            if self.left is None:
                self.left = TreeNode(val)
            else:
                self.left.insert(val)
        # insert to the right
        else:  # self.val < val
            if self.right is None:
                self.right = TreeNode(val)
            else:
                self.right.insert(val)

    def __eq__(self, other: object) -> bool:
        # use recursive value equals or equality:)
        if not isinstance(other, TreeNode):
            return NotImplemented
        return self.val == other.val and self.left == other.left and self.right == self.right

    # BST display code source
    # https://stackoverflow.com/questions/34012886/print-binary-tree-level-by-level-in-python
    def __str__(self) -> str:
        lines, *_ = self._display_aux()
        return "\n".join(lines)

    def _display_aux(self) -> tuple[list[str], int, int, int]:
        """Returns list of strings, width, height, and horizontal coordinate of the root."""
        # No child.
        if self.right is None and self.left is None:
            line = "%s" % self.val
            width = len(line)
            height = 1
            middle = width // 2
            return [line], width, height, middle

        # Only left child.
        if self.right is None:
            lines, n, p, x = self.left._display_aux()  # type: ignore
            s = "%s" % self.val
            u = len(s)
            first_line = (x + 1) * " " + (n - x - 1) * "_" + s
            second_line = x * " " + "/" + (n - x - 1 + u) * " "
            shifted_lines = [line + u * " " for line in lines]
            return [first_line, second_line] + shifted_lines, n + u, p + 2, n + u // 2

        # Only right child.
        if self.left is None:
            lines, n, p, x = self.right._display_aux()
            s = "%s" % self.val
            u = len(s)
            first_line = s + x * "_" + (n - x) * " "
            second_line = (u + x) * " " + "\\" + (n - x - 1) * " "
            shifted_lines = [u * " " + line for line in lines]
            return [first_line, second_line] + shifted_lines, n + u, p + 2, u // 2

        # Two children.
        left, n, p, x = self.left._display_aux()
        right, m, q, y = self.right._display_aux()
        s = "%s" % self.val
        u = len(s)
        first_line = (x + 1) * " " + (n - x - 1) * "_" + s + y * "_" + (m - y) * " "
        second_line = x * " " + "/" + (n - x - 1 + u + y) * " " + "\\" + (m - y - 1) * " "
        if p < q:
            left += [n * " "] * (q - p)
        elif q < p:
            right += [m * " "] * (p - q)
        zipped_lines = zip(left, right)
        lines = [first_line, second_line] + [a + u * " " + b for a, b in zipped_lines]
        return lines, n + m + u, max(p, q) + 2, n + u // 2


def build_bt(node_lst: list[Any]) -> Optional[TreeNode]:
    # empty tree
    n = len(node_lst)
    if not n:
        return None
    root = TreeNode(node_lst[0])

    q: deque[TreeNode] = deque([root])

    index = 1
    while index < n:
        cur = q.popleft()

        # index + 1 should exist by the structure of the input (i.e. if the last
        # leaf in empty should give null and not omit it)
        left = node_lst[index]
        right = node_lst[index + 1] if index + 1 < n else None

        # check None, if not null add to tree and put on queue for processing
        if left is not None:
            cur.left = TreeNode(left)
            q.append(cur.left)
        if right is not None:
            cur.right = TreeNode(right)
            q.append(cur.right)

        index += 2  # processed two more nodes

    return root

File: cpp/utils/binary_tree.cpp
/**
 * Original Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#include "binary_tree.h"
#include <queue>

using std::string;
using std::vector;
// Enable to print vectors just by calling its name

std::tuple<std::vector<string>, int, int, int> TreeNode::display_aux() {
    int width, height, middle;
    string s = std::to_string(val);
    int u = s.length();

    if (!right && !left) {
        std::vector<string> lines;
        lines.emplace_back(s);
        width = u;
        height = 1;
        middle = width / 2;
        return std::make_tuple(lines, width, height, middle);
    }

    if (!right) {
        auto [lines, n, p, x] = left->display_aux();
        string first_line = string(x + 1, ' ') + string(n - x - 1, '_') + s;
        string second_line = string(x, ' ') + '/' + string(n - x - 1 + u, ' ');

        std::vector<string> shifted_lines;
        for (auto &&line : lines)
            shifted_lines.emplace_back(line + string(u, ' '));

        shifted_lines.emplace(shifted_lines.begin(), second_line);
        shifted_lines.emplace(shifted_lines.begin(), first_line);
        return std::make_tuple(shifted_lines, n + u, p + 2, n + u / 2);
    }

    if (!left) {
        auto [lines, n, p, x] = right->display_aux();
        string first_line = s + string(x, '_') + string(n - x, ' ');
        string second_line = string(u + x, ' ') + '\\' + string(n - x - 1, ' ');

        std::vector<string> shifted_lines;
        for (auto &&line : lines)
            shifted_lines.emplace_back(string(u, ' ') + line);

        shifted_lines.emplace(shifted_lines.begin(), second_line);
        shifted_lines.emplace(shifted_lines.begin(), first_line);
        return std::make_tuple(shifted_lines, n + u, p + 2, u / 2);
    }

    std::vector<string> lines;
    auto [left_lines, n, p, x] = left->display_aux();
    auto [right_lines, m, q, y] = right->display_aux();
    string first_line = string(x + 1, ' ') + string(n - x - 1, '_') + s + string(y, '_') + string(m - y, ' ');
    string second_line = string(x, ' ') + '/' + string(n - x - 1 + u + y, ' ') + "\\" + string(m - y - 1, ' ');
    if (p < q)
        for (size_t i = 0; i < q - p; i++)
            left_lines.emplace_back(string(n, ' '));
    else if (q < p)
        for (size_t i = 0; i < p - q; i++)
            right_lines.emplace_back(string(m, ' '));

    int l = std::min(left_lines.size(), right_lines.size());
    lines.reserve(l);
    for (size_t i = 0; i < l; i++)
        lines.emplace_back(left_lines[i] + string(u, ' ') + right_lines[i]);
    lines.emplace(lines.begin(), second_line);
    lines.emplace(lines.begin(), first_line);
    return std::make_tuple(lines, n + m + u, std::max(p, q) + 2, n + u / 2);
}

std::vector<int *> parse_leetcode_bt(string s, char del) {
    std::vector<int *> node_list;
    std::stringstream ss(s);
    string word;
    int *pt;
    while (!ss.eof()) {
        getline(ss, word, del);
        if (word == "null")
            node_list.emplace_back(nullptr);
        else
            node_list.emplace_back(new int(stoi(word)));
    }
    return node_list;
}

/**
 * @brief Create a TreeNode from LeetCode list convention format of int or null_ptr
 *
 * @param node_list
 * @return TreeNode *
 */
TreeNode *build_bt(std::vector<int *> node_list) {
    auto n = node_list.size();
    if (n == 0)
        return nullptr;
    auto root = new TreeNode(*node_list[0]);

    std::queue<TreeNode *> q;
    q.push(root);
    size_t index = 1;

    while (index < n) {
        auto cur = q.front();
        q.pop();
        int *left_val_pt = node_list[index];
        int *right_val_pt = (index + 1 < n) ? node_list[index + 1] : nullptr;

        if (left_val_pt) {
            auto left_node = new TreeNode(*left_val_pt);
            cur->left = left_node;
            q.push(cur->left);
        }
        if (right_val_pt) {
            auto right_node = new TreeNode(*right_val_pt);
            cur->right = right_node;
            q.push(cur->right);
        }
        index += 2;
    }
    return root;
}

TreeNode *build_bt(string leetcode_bt) { return build_bt(parse_leetcode_bt(leetcode_bt)); }

std::ostream &operator<<(std::ostream &output, TreeNode &root) {
    auto [lines, _1, _2, _3] = root.display_aux();
    for (auto &&line : lines)
        output << line << "\n";
    return output;
}

std::ostream &operator<<(std::ostream &output, TreeNode *&root) {
    auto [lines, _1, _2, _3] = root->display_aux();
    for (auto &&line : lines)
        output << line << "\n";
    return output;
}